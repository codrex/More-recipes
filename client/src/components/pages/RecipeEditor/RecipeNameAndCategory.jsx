import React from 'react';
import { Field, reduxForm, formValueSelector } from 'redux-form';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import { Form, Select, Input } from '../../common/FormElements';
import { recipeName } from '../../../utils/validator';
import { CATEGORIES } from '../../../constants';
import {
  updateName,
  updateCategory,
  clearValidationError as clearError
} from '../../../actions/recipeActions';

/**
 * RecipeNameAndCategory component
 * @param {object} props
 * @return {React} react element
 */
let RecipeNameAndCategory = (props) => {
  const {
    name,
    recipe,
    validationError,
    clearValidationError
  } = props;
  const {
    name: nameError,
    category: categoryError
  } = validationError;
  return (
    <div>
      <h4 className="items-header-text">
        Recipe name and category
      </h4>
      <Form style={{ zIndex: 4 }}>
        <Field
          component={Input}
          name="name"
          type="text"
          id="recipeName"
          placeholder="Enter recipe name"
          className="add-recipe-input"
          validate={value => recipeName(value, 'recipe name')}
          onBlur={() => props.updateName(name)}
          onFocus={() => clearValidationError({
            name: ''
          })}
          externalError={{
            touched: true,
            error: nameError
          }}
        />
        <Select
          name="category"
          type="text"
          id="category"
          placeholder="Select recipe category"
          className="add-recipe-input"
          options={CATEGORIES}
          value={recipe.category}
          handleChange={({ value }) => {
            props.updateCategory(value);
            clearValidationError({
              category: ''
            });
          }}
          externalError={{
            touched: true,
            error: categoryError
          }}
        />
      </Form>
    </div>
  );
};

RecipeNameAndCategory.defaultProps = {
  recipeCategory: '',
  name: ''
};

RecipeNameAndCategory.propTypes = {
  name: PropTypes.string,
  clearValidationError: PropTypes.func.isRequired,
  recipe: PropTypes.objectOf(PropTypes.shape).isRequired,
  updateName: PropTypes.func.isRequired,
  updateCategory: PropTypes.func.isRequired,
  validationError: PropTypes.objectOf(PropTypes.shape).isRequired,
};

// For testing purpose
const PureRecipeNameAndCategory = RecipeNameAndCategory;
export { PureRecipeNameAndCategory };

RecipeNameAndCategory = reduxForm({
  form: 'recipeNameAndCategory',
})(RecipeNameAndCategory);

const selector = formValueSelector('recipeNameAndCategory');

const mapStateToProps = state => ({
  recipe: state.recipe,
  initialValues: { name: state.recipe.name },
  name: selector(state, 'name'),
  validationError: state.recipeValidationError,
});

export default connect(mapStateToProps, {
  updateCategory,
  updateName,
  clearValidationError: clearError
})(RecipeNameAndCategory);
