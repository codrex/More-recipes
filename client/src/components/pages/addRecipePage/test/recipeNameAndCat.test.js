import React from 'react';
import RecipeNameAndCategory from '../recipeNameAndCategory/recipeNameAndCategory';
import { Provider } from 'react-redux';
import toJSON from "enzyme-to-json"
import configureStore from '../../../../store/configStore';

const props = {
  handleSubmit: jest.fn(),
  actions: {
    updateStore: jest.fn(),
    postRecipe: jest.fn(),
    modifyRecipe: jest.fn(),
  },
  modify: '',
  create: '',
  loading: false,
}
describe('Recipe name and category component', () => {
  test('render as expected', () => {
    const component = shallow(
      <Provider store={configureStore()}>
        <RecipeNameAndCategory {...props} />
      </Provider>);
    const tree = toJSON(component);
    expect(tree).toMatchSnapshot();
  });
});

