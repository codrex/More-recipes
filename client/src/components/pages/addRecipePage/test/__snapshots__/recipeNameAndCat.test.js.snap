// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Recipe name and category component render as expected 1`] = `
<div>
  <h4
    className="lead items-header-text"
  >
    Name and category
  </h4>
  <form
    action={undefined}
    className="col-12"
    method={undefined}
    onSubmit={[Function]}
  >
    <div
      className="form-group"
    >
      <input
        aria-describedby="recipeName"
        autoComplete="false"
        className="form-control text-input add-recipe-input"
        id="recipeName"
        name="recipeName"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder="Enter recipe name"
        type="text"
        value=""
      />
    </div>
    <div
      className="form-group"
    >
      <input
        aria-describedby="category"
        autoComplete="false"
        className="form-control text-input add-recipe-input"
        id="category"
        name="category"
        onBlur={[Function]}
        onChange={[Function]}
        onDragStart={[Function]}
        onDrop={[Function]}
        onFocus={[Function]}
        placeholder="Enter recipe category"
        type="text"
        value=""
      />
    </div>
    <button
      className="btn text-uppercase w-100 btn-secondary"
      disabled={false}
      id="submit"
      type="submit"
    >
      create recipe
    </button>
  </form>
</div>
`;
